subprojects/model-core/src/test/groovy/org/gradle/api/internal/provider/MapPropertySpec.groovy

```diff
@@ -1063,6 +1063,109 @@ The value of this property is derived from: <source>""")
         return new DefaultProperty<String>(host, String)
     }
 
+    def "runs side effect when calling '#getter' on property to which providers were added via 'put'"() {
+        when:
+        def sideEffect1 = Mock(ValueSupplier.SideEffect)
+        def sideEffect2 = Mock(ValueSupplier.SideEffect)
+        property.put("some key", Providers.of("some value").withSideEffect(sideEffect1))
+        property.put("other key", Providers.of("other value").withSideEffect(sideEffect2))
+
+        property.calculateValue(ValueSupplier.ValueConsumer.IgnoreUnsafeRead)
+        property.calculateExecutionTimeValue()
+
+        then:
+        0 * _ // no side effects until values are unpacked
+
+        when:
+        def unpackedValue = getter(property, getter, ["yet another key": "yet another value"])
+
+        then:
+        unpackedValue == ["some key": "some value", "other key": "other value"]
+        1 * sideEffect1.execute("some value")
+
+        then: // ensure ordering
+        1 * sideEffect2.execute("other value")
+        0 * _
+
+        where:
+        getter      | _
+        "get"       | _
+        "getOrNull" | _
+        "getOrElse" | _
+    }
+
+    def "runs side effect when calling '#getter' on property to which providers were added via 'putAll'"() {
+        when:
+        def sideEffect = Mock(ValueSupplier.SideEffect)
+        property.putAll(Providers.of(someValue()).withSideEffect(sideEffect))
+
+        property.calculateValue(ValueSupplier.ValueConsumer.IgnoreUnsafeRead)
+        property.calculateExecutionTimeValue()
+
+        then:
+        0 * _ // no side effects until values are unpacked
+
+        when:
+        def unpackedValue = getter(property, getter, ["yet another key": "yet another value"])
+
+        then:
+        unpackedValue == someValue()
+        1 * sideEffect.execute(someValue())
+        0 * _
+
+        where:
+        getter      | _
+        "get"       | _
+        "getOrNull" | _
+        "getOrElse" | _
+    }
+
+    def "runs only own side effect when calling '#getter' on property's 'keySet'"() {
+        given:
+        def ownSideEffect = Mock(ValueSupplier.SideEffect)
+        def valueSideEffect = Mock(ValueSupplier.SideEffect)
+        property.withSideEffect(ownSideEffect)
+
+        when:
+        property.put("some key", Providers.of("some value").withSideEffect(valueSideEffect))
+        def keySetValue = property.keySet().get()
+
+        then:
+        keySetValue == ["some key"].toSet()
+        0 * ownSideEffect.execute(["some key": "some value"])
```

Comment:
Shouldn't this run the side effect here?

Reply:
Oh, I actually forgot to delete this test, because it does not make sense.

As of now, there is no such thing as "own side effect" on a MapProperty or a CollectionProperty. Method `withSideEffect` always returns a new Provider that would wrap around the parent. When this happens, the return provider is not a MapProperty or a CollectionProperty anymore and therefore you can't access any method specific to them.

The bug in this test is that I simply do that and the actual returned provider with `ownSideEffect` is discarded:
```
        property.withSideEffect(ownSideEffect)
```

Reply:
Removed the test